---
title: Introduction to NCBI
author: "Nathaniel Maki"
date: "May 17th, 2020"
---
## Learning Objectives
* Acquire a basic understanding of the NCBI website, its resources, and its most commonly used utilities/tools
* More specifically, become familier with Pubmed, the Gene Expression Omnibus (GEO), BLAST, and briefly the Sequence Read Archive (SRA)
* Exposure to future and upcoming tools/resources under active development, such as NCBI Datasets, NCBI Virus Repository, and BLAST+

## So, what exactly is the NCBI?
* National Center for Biotechnology Information
* Three main focuses:
* Maintain a vast repository of data, generated by researchers from all over the world
  * GEO, SRA, etc
* Develop tools, utilities, and software for procuring available data, and for performing analysis
  * BLAST, SRA-toolkit, RAPT
* Provide a web interface for interacting with, manipulating, and querying said data

In addition the the above, they also produce training materials to assist in orienting students, fledgling researchers, and seasoned PIs to the tools/resources they offer. 

The following [Link](https://www.ncbi.nlm.nih.gov/home/learn/) will take you to the NCBI education page.

## Front Page, Explained
<img src="./ncbi_images/ncbi_home.png" width="800">

The front page of NCBI acts as a hub of sorts, with the main source of navigation being the search bar, and dropdown menu beside it. At the top left of the page are quick links to resources available, as well as grouped tutorials covering common use cases.

<img src="./ncbi_images/ncbi_resources_dropdown.png" width="400">

Selecting BLAST will take you to the tools' homepage:

<img src="./ncbi_images/ncbi_blast.png" width="800">

By contrast, the dropdown menu:

<img src="./ncbi_images/ncbi_howto_dropdown.png" width="400">

Selecting Data and Software has some useful tutorials right off the bat:

<img src="./ncbi_images/howto_expanded.png" width="800">

Clicking the how-to guide on downloading the complete genome of an organism takes you to the FTP intro page:

<img src="./ncbi_images/howto_ftp_teaser.png" width="800">

Now, back to the front page:

<img src="./ncbi_images/ncbi_home.png" width="800">

The resources column on the left brings you to a more comprehensive list of available tools (for say, Data + Software). It's a bit confusing having two sections of the main page refering to resources, and is something that's being unified when they finish revamping the site.

On the right hand side are popular resources within the NCBI site. You'll notice that a few of the mentioned above subjects are included in this list.

Content in the center is comprised of links to the main functions of th NCBI site
* Submit: NCBI Submission Portal, for uploading data to their online repositories
* Download: use FTP to bulk download data from NCBI directly to your local machine
* Learn: location for educational materials/tutorials/etc
* Develop: tools and resources for software/web developers, link to NCBI Github repo
* Analyze: find a tool for manipulating, aligning, visualizing, and evaluating biological data
* Research: current tools and basic research topics being worked on by the NCBI Compuational Biology branch

## Core of NCBI: The Entrez Query
Entrez acts as the primary method of executing text queries and retrieval.
It integrates the entire PubMed library of related biomedical literature with a myriad of other relevent literature and moleculare databases. These libraries include:
* DNA + protein sequence + structure
* gene
* genome
* genetic variation
* gene expression

It can be a bit daunting to use at first, and general searches can leave you with an overwhelming number of results. I'll be showing you how to structure your queries in such a way, as to take advantage of Entrez's flexible and powerful filter functions.

For every database that is available, there are particular and specialized fields that can be used to help refine your results. 

* It's worth mentioning here that if you haven't yet, creating an NCBI account greatly increases the flexibility of the tools and resources at your dispolal through this site.
  * A simple example being in working with NCBI's Entrez Programming Utilities:
  * With a verified account, your database calls are not throttled, and you're afforded more compute than if you were to remain anonomyous.

Additionally, Entrez integrates relevent data between databases. This means that if you are examining a Gene Expression Omnibus (GEO) profile of some experiment, a link to the appropriate SRA page will be included (and vice versa)

### Select Entrez Databases
To be specific, Entrez encompasses 39 molecular and literature databases. Of course, we're not going to be covering them all, but here are some of what I consider the more relevant ones:

* GEO Datasets
* GEO Profiles
* SRA
* PubMed + PubMed Central

### GEO Datasets
* Stores curated gene expression + molecular abundance data put together by NCBI
* Sourced from the GEO repository of microarray data
* Made up of original submitted records, including series, samples + platforms used
* Records contain resources and tools, such as clustering utilities and differential expression searches

#### GEO Dataset Results page:

<img src="./ncbi_images/GEO_datasets_results.png" width="800">

* Title of each individual dataset lists either the Dataset (GDS), Series (GSE), or Platform (GPL) accession number
* Followed by Title, and Organism in brackets
* Underneath Accession number are quick links to the GEO Profiles, PMC text, and tools for analysis
* Clicking the main title brings you to the Dataset Record

#### GEO Dataset Records page:

<img src="./ncbi_images/GEO_dataset_record.png" width="800">

* Main record contains descriptive information about the dataset (organism, platform, sample values + number)
* Next to the record, the Expression Profiles link takes you to the GEO Profiles page
* Cluster Analysis to the right brings you to the requisite heatmap
* Under that are options to download various files containg additional Dataset information, experimental variable subsets, etc
* At the bottom are additional data analysis tools for finding genes, comparing sets of samples, generating heatmaps, and examining experimental design + value distributions.

### GEO Profiles
* A derivative of Geo Datasets
* Stores *individual* curated gene expression + molecular abundance profiles
* Assembled and sourced from GEO microarray data
* Queries based upon gene annotation / profile characteristics

#### GEO Profile Results page:

<img src="./ncbi_images/GEO_profile_results.png" width="800">

* Lists Dataset accession (GDS) from where the profile is derived from
* Platform accession (GPL) that the Dataset is based on + unique ID from record
* Organism
* Additionally, contains:
  * Annotation: gene symbol, full name + aliases
  * Reporter: Original sequence reporter(s) taken from the Platform record supplied by submitter
  * Experiment: Dataest from where the profile comes from

#### GEO Profile Chart page:

<img src="./ncbi_images/GEO_profile_record.png" width="600">

* By clicking on the little chart thumbnail, the full profile details are enlarged
* This expanded chart include more comprehensive details, such as expression values and experimental design
* Explaining the graph:
  * Each column you see represents the expression measurement that was extracted from the VALUE column of the Sample record
  * Original Sample accessions are listed below (GSM)
  * Something to note: The values associated with each accession have undergone processing and normalization that is not uniform across accessions.
  * Because of this, the values should be considered arbitrary, and direct comparisons between different Datasets may not be accurate.
  * The square represent rank order of expression measurements, and indicate where the expression of that gene falls in comparison to all other genes on an array.

#### GEO Sample Accession

<img src="./ncbi_images/GEO_sample_accessions.png" width="800">

* Selecting a Sample in the Profile Chart brings you to it's appropriate GEO Accession page
* Here, you can view tons of relevent information in regards to the singular Sample including: 
  * Sample type
  * Source name
  * Organism
  * Various characteritics
  * Protocols used for Treatment, Growth, Extraction, etc
* Also includes the Platfrom ID (GPL) and Series ID at the bottom
* You also have the option to download raw Sample specific CEL data, either through a web browser, or through FTP

### SRA
* Contains sequencing data from *all* NGS platforms (Illumna, Roche, etc)
* Largest publicaly available repo of HTS data
* Stores raw sequencing data + alignment information
* Accompanied by SRA-toolkit suite for acquisition + dumping of data from SRA database

### PubMed + PubMed Central
* Biomedical Literature database
* Includes citations + abstracts, from journals such as Nature, Medline, etc
* Wealth of information, however... often paywalled
* PubMed "Vanilla" acts mainly as a reference source, with just the above, citations + abstracts
* PubMed Central however, contains full-text versions of articles for free
* If available, PubMed will link to PMC for appropriate article

## Tools and Additional Resources

## BLAST
* Searches for areas of similarity between biological sequences
* Compares various nucleotide or protein sequences to sequence databases, and calculates statistical significance of matches
* Can determine functional + evolutionary reltionships between sequences
* Identify members of gene families
* Can be ran through a web browser, referenced in external code, installed locally, or executed in a cloud environment through BLAST+ (AWS, Azure, etc)
* There also exist many searches that have been tweaked for specific use cases, such as: 
  * CD-search, which locates conserved domains in submitted sequence
  * Primer-BLAST, which allows you to design primers that are specific to a PCR template (really neat)

## NCBI Datasets (Beta)

<img src="./ncbi_images/ncbi_datasets.png" width="800">

* Resource that NCBI is developing for locating and constructing datasets
* Allows for bulk downloading of genome sequence as well as matching annotation
* For now, web interface only works with Eukaryotic organisms
* CLI is needed for all other organisms
* The site is a bit sparse at the moment, and selecting say, *Mus musculus*, will present you with only a couple reference genomes, and a single available annotation
* As this resource matures, expect additiotial usefulness to follow

## NCBI Virus (Beta)

<img src="./ncbi_images/ncbi_virus.png" width="800">

* Acts as a central hub of sorts, for viral sequence data originating from GenBank, Refseq, and other NCBI repositories
* Includes a really neat visual aid for the taxonomy of viruses, discriminating between hosts
* Especially relevent today, you can find a link under "Find Data" called "Up-to-date-SARS-Cov2" that brings you to a hub page
* Here you can browse the vast amount of nucleotide and protein data, locate PubMed and PubMed Central pages, perform in-place alignments, and construct a phylogenetic tree